### PR TITLE
Apple: Do not specify an SDK version in `rlib` object files

### PR DESCRIPTION
**Pull request by madsmtm**
This was added in https://github.com/rust-lang/rust/pull/114114, but is unnecessary, since it ends up being overwritten when linking anyhow, and it feels wrong to embed some arbitrary SDK version in here. The object files produced by LLVM also do not set this, and the tooling shows `n/a` when it's `0`, so it seems to genuinely be optional in object files.

I've also added a test for the different places the SDK version shows up, and documented a bit more in the code how SDK versions work.

See https://github.com/rust-lang/rust/issues/129432 for the bigger picture.

Tested with (excludes the same few targets as in https://github.com/rust-lang/rust/pull/130435):
```console
./x test tests/run-make/apple-sdk-version --target aarch64-apple-darwin,aarch64-apple-ios,aarch64-apple-ios-macabi,aarch64-apple-ios-sim,aarch64-apple-tvos,aarch64-apple-tvos-sim,aarch64-apple-visionos,aarch64-apple-visionos-sim,aarch64-apple-watchos,aarch64-apple-watchos-sim,arm64_32-apple-watchos,armv7k-apple-watchos,armv7s-apple-ios,x86_64-apple-darwin,x86_64-apple-ios,x86_64-apple-ios-macabi,x86_64-apple-tvos,x86_64-apple-watchos-sim,x86_64h-apple-darwin
IPHONEOS_DEPLOYMENT_TARGET=10.0 ./x test tests/run-make/apple-sdk-version --target=i386-apple-ios
```

CC @BlackHoleFox, you [originally commented on these values](https://github.com/rust-lang/rust/pull/114114#discussion_r1300599445).

@rustbot label O-apple

### CODE REVIEW

File: tests/run-make/apple-sdk-version/rmake.rs [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        if crate_type == "dylib" && apple_os() == "watchos" {
+            continue;
+        }
+
+        // Test with clang
+        let file_name = format!("foo_cc{file_ext}");
+        rustc()
+            .target(target())
+            .crate_type("bin")
+            .arg("-Clinker-flavor=gcc")
```
Comment (jieyouxu) [inline tests/run-make/apple-sdk-version/rmake.rs:75]: Question: is this comment or arg accurate, the comment says clang, but linker-flavor is gcc?

**Review by jieyouxu — APPROVED** [review]
Thanks, the changes look reasonable to me but I'll wait for other apple target maintainers to double-check. I have one question regarding the test.

**Comment by jieyouxu** [comment]
Should this get relnotes? Even if it doesn't technically change how object file SDK version is treated if I understand this correctly.

File: tests/run-make/apple-sdk-version/rmake.rs (hunk 1 above)
Comment (madsmtm) [reply tests/run-make/apple-sdk-version/rmake.rs:75]: The comment is mostly accurate, it tests with `cc`, which is basically always Clang on Apple platforms, it's just how [the `linker-flavor`](https://doc.rust-lang.org/rustc/codegen-options/index.html#linker-flavor) option works.

**Comment by madsmtm** [comment]
> Thanks, the changes look reasonable to me but I'll wait for other apple target maintainers to double-check. I have one question regarding the test.

I'll tag @thomcc as well then.

> Should this get relnotes? Even if it doesn't technically change how object file SDK version is treated if I understand this correctly.

This is only used inside `create_object_file`, and that, in turn, is only used for constructing `rlib`s. So no, it doesn't make sense to relnote, the difference is only observable if you manually unarchive the `rlib` (and the format of that is explicitly unspecified).

**Review by BlackHoleFox — APPROVED** [review]
Nice PR, everything looks fine to me and more test coverage too.

I did indeed comment on the original values because Magic Numbers are odd, but it was harmless in removing a linker warning with no actual cost beyond a strangeness factor. I didn't dig in very far (should have looked at LLVM like you did). Since it turns out the SDK version is actually optional at this stage of linking, there doesn't appear to be any good reason left for `rustc` to carry that added strange.

**Comment by jieyouxu** [comment]
Thanks @BlackHoleFox . In that case,

@bors r+ rollup

**Pull request merged**